### PR TITLE
xorg* (new formulae)

### PR DESCRIPTION
**Pull request by iMichka**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

-----

These formulae were adapted from https://github.com/Linuxbrew/homebrew-xorg, but only the ones from the xorg.rb meta-package.

This is a draft PR to prepare their migration. Some tests are missing and need to be written.

Once this PR is green, I can split each formula in separate commits. I would like to keep this under one single PR though and merge this in a single step.

### CODE REVIEW

**Comment by maxim-belkin** [comment]
@sjackman requested that these formulae are migrated one by one. I will submit PRs when they are ready. Feel free to submit PRs to linuxbrew/xorg to add missing tests.

**Comment by iMichka** [comment]
I think we need to migrate them all at once. Because else we will need to revision bump xorg 42 times, and we will need to rebottle 42 times all the xorg-dependent formulae in linuxbrew-core too.

If we merge everything at once, I can unbottle stuff on linuxbrew-core, delete the xorg formula from the xorg-tap, merge homebrew-core to linuxbrew-core, and rebuild everything.
It is a tricky operation which might last a few hours, I already did it of mesa with success. I would like to avoid to have to do it 42 times :)

**Comment by maxim-belkin** [comment]
My bad!!! Shaun suggested that these are migrated all at once! So, it's all good. Sorry for the confusion. If you're happy with the tests/etc -- feel free to merge/migrate.

**Comment by iMichka** [comment]
No issue. As I already experimented with this before, I think this is the best solution.

I'll try to fix as many issues here until CI is green, and ping you for a review.

**Comment by jonchang** [comment]
brew test --verbose libx11

```
==> /usr/bin/clang test.c -o test -I/usr/local/Cellar/libx11/1.6.9/include -lX11
ld: library not found for -lX11
clang: error: linker command failed with exit code 1 (use -v to see invocation)
Error: libx11: failed
```

brew test --verbose libsm

```
==> /usr/bin/clang test.c -o test -ldl -I/usr/local/Cellar/libsm/1.2.3/include -L/usr/local/Cellar/libsm/1.2.3/lib -lSM
==> ./test
dlopen(/usr/local/Cellar/libsm/1.2.3/lib/libSM.so, 1): image not found
Error: libsm: failed
```

brew linkage --test libxpm


```
Broken dependencies:
  /usr/local/opt/gettext/lib/libintl.8.dylib (gettext)
```


brew install --verbose --build-bottle libpciaccess

```
libtool: compile:  clang -DHAVE_CONFIG_H -I. -I.. -I../include -Wall -Wpointer-arith -Wmissing-declarations -Wformat=2 -Wstrict-prototypes -Wmissing-prototypes -Wnested-externs -Wbad-function-cast -Wold-style-definition -Wdeclaration-after-statement -Wunused -Wuninitialized -Wshadow -Wmissing-noreturn -Wmissing-format-attribute -Wredundant-decls -Werror=implicit -Werror=nonnull -Werror=init-self -Werror=main -Werror=missing-braces -Werror=sequence-point -Werror=return-type -Werror=trigraphs -Werror=array-bounds -Werror=write-strings -Werror=address -Werror=int-to-pointer-cast -Werror=pointer-to-int-cast -fno-strict-aliasing -g -O2 -c common_capability.c  -fno-common -DPIC -o .libs/common_capability.o
common_interface.c:75:10: fatal error: 'sys/endian.h' file not found
common_init.c:73:3: error: "Unsupported OS"
#include <sys/endian.h>
         ^~~~~~~~~~~~~~
# error "Unsupported OS"
  ^
1 error generated.
make[2]: *** [common_init.lo] Error 1
make[2]: *** Waiting for unfinished jobs....
1 error generated.
make[2]: *** [common_interface.lo] Error 1
```

brew test --verbose libxfont

```
==> ldd /usr/local/Cellar/libxfont/1.5.4/lib/libXfont.so
Error: libxfont: failed
An exception occurred within a child process:
  Errno::ENOENT: No such file or directory - ldd
```

**Comment by samford** [comment]
While working on the livecheck situation for these formulae, I noticed that the x.org archive URLs are redirecting, so you may want to address it in these formulae.

For example:

* libdmx: https://ftp.x.org/pub/individual/lib/libdmx-1.1.4.tar.bz2 -> https://ftp.x.org/archive/individual/lib/libdmx-1.1.4.tar.bz2
* libpciaccess: https://www.x.org/pub/individual/lib/libpciaccess-0.16.tar.bz2 -> https://www.x.org/archive/individual/lib/libpciaccess-0.16.tar.bz2

There's also something to be said about standardizing on ftp.x.org or www.x.org URLs. These URLs seem to be interchangeable between the `ftp` and `www` subdomains.

**Comment by iMichka** [comment]
- Removed livechecks
- Aligned urls
- Fixed libxpm linkage
- Fixed libx11 test

Lets run this a second time. There are also a lot of tests still missing.

**Comment by samford** [comment]
```
==> brew audit --skip-style xdpyinfo --new-formula
==> FAILED
Error: 1 problem in 1 formula detected
xdpyinfo:
  * Stable: The URL https://www.x.org/archive/individual/lib/xdpyinfo-1.3.2.tar.bz2 is not reachable (HTTP status code 404)
```

The URL for `xdpyinfo` should be: https://www.x.org/archive/individual/app/xdpyinfo-1.3.2.tar.bz2

**Comment by SMillerDev** [comment]
```
brew test --verbose libsm
16
brew install --verbose --build-bottle libpciaccess
17
brew test --verbose libxfont
18
brew fetch --retry xdpyinfo --build-bottle --force
19
brew install --verbose --build-bottle xdpyinfo
20
brew audit --skip-style xdpyinfo --new-formula
```

**Comment by iMichka** [comment]
- deleted: libpciaccess.rb: is for Linux only

- modified:   libsm.rb
- modified:   util-macros.rb
- modified:   xcb-proto.rb
- modified:   xdpyinfo.rb
- modified:   xorgproto.rb
- modified:   xtrans.rb

Fixed urls and added tests

**Comment by iMichka** [comment]
What is our policy for packages that contain only headers / macros.
I see that we vendored some of these packages, like xproto and xorg-macros in the `makedepend`, but this formula predates the xorg inclusion.

If we choose to vendor the proto packages and xorg-macros one, this will mean quite a bunch of vendoring (at least on linux). So I would prefer to have plain formulae for all of these.

**Comment by SMillerDev** [comment]
> What is our policy for packages that contain only headers / macros.

Go for it.

**Comment by iMichka** [comment]
This is now ready for review.

One question is about the license: https://www.x.org/archive/current/doc/xorg-docs/License.html
I tried to have a look, some packages have 36 sublicences, almost 1 per file. It's quite a mess. So I added MIT everywhere because this is what their website tells me. Still, these are variants of the MIT License, so I am unsure how to handle this.

File: Formula/font-util.rb
Comment (SMillerDev) [inline Formula/font-util.rb:33]: Can we assert anything on/after this?

File: Formula/libdmx.rb
Comment (SMillerDev) [inline Formula/libdmx.rb:38]: Printing some text might be easier

File: Formula/libpthread-stubs.rb
Comment (SMillerDev) [inline Formula/libpthread-stubs.rb:16]: Can't you check them with pkg-config?

**Comment by Bo98** [comment]
> some packages have 36 sublicences, almost 1 per file. It's quite a mess. So I added MIT everywhere because this is what their website tells me. Still, these are variants of the MIT License, so I am unsure how to handle this.

Most MIT derivates should be ok to group at least. I think Debian do this.

There are some derivates explicitly defined by SPDX like SGI though. We however don't have a mechanism to specify multiple licenses where they _all_ apply. This question has come up a few times now.

**Comment by maxim-belkin** [comment]
@iMichka, Libxcb test is out-of-date

**Comment by jonchang** [comment]
@imichka CI for this is green. I can split the commits and push bottles if you think this is ready.

**Comment by iMichka** [comment]
I just came back from holidays. I need to finish this, I had some more changes to push, and I think there are some comments above that I need to address. I’ll do the splitting myself when it is done.

File: Formula/libdmx.rb
Comment (iMichka) [reply Formula/libdmx.rb:38]: I'll make a more general answer to this below, as this impacts most of the tests.

**Comment by iMichka** [comment]
This is ready for a new round of reviews.

@SMillerDev you asked if I could print anything in some tests. I think I could but I did not because:
- Most tests are just compiled and not executed: on Linux there is no X windowing system so I'll have to disable these test on Linux, adding a bunch of if/elses in the test blocks later on.
- I did not research what to print and I have no clue about the xorg API: I just quickly copy pasted some stuff from their header files to make sure I can compile the library. Yes, I am lazy and had not much motivation to write around 40 real-world tests.
- I do not expect all these formulae to be usable as-is on Mac: there might be runtime issues I am not aware of, and I do not expect these to be a drop-in replacement for Xquartz: I'll leave that to someone who has more time than me. Adding these formulae is a first step to get rid of Xquartz: making it work is a second one.

**Review by SMillerDev — DISMISSED** [review]

**Comment by iMichka** [comment]
Just FYI: I'll merge this myself when I am ready and have a few hours in front of me. Merging this involves a bunch of bottle rebuilds in linuxbrew-core and homebrew-xorg; so I prefer to push this myself.

**Pull request closed without merging**
